cjdns-snode/src/server/service.rs

```diff
@@ -44,38 +50,63 @@ async fn do_service(server: Arc<Server>) -> Result<(), Error> {
         mktime(0xffffffffffffffff),
         ipv6,
         None,
-    )?;
+    ).expect("internal error: unknown encoding scheme");
     server.mut_state.lock().self_node = Some(Arc::new(self_node));
 
     debug!("Got selfNode");
 
     // Starting to sniff traffic
     let mut sniffer = Sniffer::sniff_traffic(cjdns, ContentType::Cjdht).await?;
 
-    // todo check connection impl after implementing new routes for test_srv
-
     // Handling subnode messages
-    loop {
-        select! {
-            msg = sniffer.receive() => {
-                match msg {
-                    Ok(msg) => {
-                        let ret_msg_opt = on_subnode_message(server.clone(), msg).await?;
-                        if let Some(ret_msg) = ret_msg_opt {
-                            sniffer.send(ret_msg, None).await?;
+    let handle_subnode_msg = async move {
+        loop {
+            select! {
+                msg = sniffer.receive() => {
+                    match msg {
+                        Ok(msg) => {
+                            let ret_msg_opt = on_subnode_message(server.clone(), msg).await?;
+                            if let Some(ret_msg) = ret_msg_opt {
+                                sniffer.send(ret_msg, None).await?;
+                            }
                         }
-                    }
 
-                    Err(err @ ReceiveError::SocketError(_)) => {
-                        return Err(err.into());
-                    }
+                        Err(err @ ReceiveError::SocketError(_)) => {
+                            return Err(err.into());
+                        }
 
-                    Err(ReceiveError::ParseError(err, data)) => {
-                        debug!("Bad message received:\n{}\n{}", hex::encode(data), anyhow!(err));
+                        Err(ReceiveError::ParseError(err, data)) => {
+                            debug!("Bad message received:\n{}\n{}", hex::encode(data), anyhow!(err));
+                        }
                     }
                 }
             }
         }
+    };
+
+    // Handlers check
+    let check_handlers_amount = async {
+        loop {
+            time::delay_for(CHECK_HANDLERS_DELAY).await;
+
+            if count_handlers().await? == 0 {
+                return Err(anyhow!("Call to UpperDistributor_listHandlers returned 0 handlers count"));
+            }
+        }
+    };
+
+    select! {
+        res = handle_subnode_msg => res,
+        res = check_handlers_amount => res
+    }
```

Comment:
Overcomplicated - two loops and two nested selects. Should be only one loop and select.